urwid/widget/columns.py

```diff
@@ -555,82 +715,228 @@ def column_widths(self, size: tuple[int] | tuple[int, int], focus: bool = False)
+        else:
+            max_height = size[1]
+
+        for idx in (*box, *box_need_height):
+            heights[idx] = max_height
+            w_h_args[idx] = (widths[idx], max_height)
+
+        return (
+            widths,
+            tuple(heights[idx] for idx in range(len(heights))),
+            tuple(w_h_args[idx] for idx in range(len(w_h_args))),
+        )
+
+    def pack(self, size: tuple[()] | tuple[int] | tuple[int, int], focus: bool = False) -> tuple[int, int]:
+        """Get packed sized for widget."""
+        if size:
+            return super().pack(size, focus)
+        widths, heights, _ = self.get_column_sizes(size, focus)
+        return (sum(widths) + self.dividechars * max(len(widths) - 1, 0), max(heights))
+
+    def render(
+        self,
+        size: tuple[()] | tuple[int] | tuple[int, int],
+        focus: bool = False,
+    ) -> SolidCanvas | CompositeCanvas:
         """
         Render columns and return canvas.
 
         :param size: see :meth:`Widget.render` for details
         :param focus: ``True`` if this widget is in focus
         :type focus: bool
         """
-        widths = self.column_widths(size, focus)
+        widths, _, size_args = self.get_column_sizes(size, focus)
 
-        box_maxrow = None
-        if len(size) == 1:
-            box_maxrow = 1
-            # two-pass mode to determine maxrow for box columns
-            for i, (mc, (w, (_t, _n, b))) in enumerate(zip(widths, self.contents)):
-                if b:
-                    continue
-                rows = w.rows((mc,), focus=focus and self.focus_position == i)
-                box_maxrow = max(box_maxrow, rows)
-
-        data = []
-        for i, (mc, (w, (_t, _n, b))) in enumerate(zip(widths, self.contents)):
+        data: list[tuple[Canvas, int, bool, int]] = []
+        for i, (width, w_size, (w, _)) in enumerate(zip(widths, size_args, self.contents)):
             # if the widget has a width of 0, hide it
-            if mc <= 0:
+            if width <= 0:
                 continue
 
-            if box_maxrow and b:
-                sub_size = (mc, box_maxrow)
-            else:
-                sub_size = (mc,) + size[1:]
-
-            canv = w.render(sub_size, focus=focus and self.focus_position == i)
-
             if i < len(widths) - 1:
-                mc += self.dividechars  # noqa: PLW2901
-            data.append((canv, i, self.focus_position == i, mc))
+                width += self.dividechars  # noqa: PLW2901
+            data.append(
+                (
+                    w.render(w_size, focus=focus and self.focus_position == i),
+                    i,
+                    self.focus_position == i,
+                    width,
+                )
+            )
 
         if not data:
-            return SolidCanvas(" ", size[0], (size[1:] + (1,))[0])
+            if size:
+                return SolidCanvas(" ", size[0], (size[1:] + (1,))[0])
+            raise ColumnsError("No data to render")
```

Comment:
not allowing to render as fixed empty widget (as flow it will be also not rendered as it uses original logic)